File: data/yaml.test.ts
import { parse } from 'yaml'
import fs from 'fs'
import { test, expect, CustomMatcherResult } from 'vitest'

const resources_path: string = 'data/resources.yml'

// Defining tests
// type resource = { title: string, url: string, tags: string[] }

interface Resource {
    title: string,
    url: string,
    tags: string[],
}

function toBeResource(toBeDetermined: any): CustomMatcherResult {
    const { title, url, description, tags } = toBeDetermined;

    // Checking that resource has 4 keys
    if (!(Object.keys(toBeDetermined).length === 4)) {
        throw new Error(`resource url ${url} has ${Object.keys(toBeDetermined).length} keys (should have 4; title, url, description, and tags)`)
    }

    // Checking that resource has the correct keys, with the correct types
    if (!(typeof url === 'string')) {
        throw new Error(`url '${url}' (for title '${title}') is not a string (is ${typeof url})`)
    } else if (!(typeof title === 'string')) {
        throw new Error(`title '${title}' (for url ${url}) is not a string`)
    } else if (!(typeof description === 'string')) {
        throw new Error(`description (for url ${url}) is not a string`)
    } else if (!(tags instanceof Array)) {
        throw new Error(`tags '${tags}' (for url ${url}) is not an array`)
    }

    // Error with empty tags
    if (tags.length === 0) {
        throw new Error(`tags '${tags}' (for url ${url}) is empty. Please give tags`)
    }

    // Checking that tags are strings
    for (let tag of toBeDetermined.tags) {
        if (!(typeof tag === 'string')) {
            throw new Error(`tag '${tag}' (for url ${url}) is not a string`)
        }
    }

    return {
        message: () => `resource '${title}' (${url}) is valid`,
        pass: true,
    }
}

expect.extend({
    toBeResource,
});

const file = fs.readFileSync(resources_path, 'utf8')
const yml_data = parse(file)

console.log(yml_data)

// check yml_data is of type resources
test('testing resources in YAML file...', () => {
    for (let resource of yml_data) {
        expect(resource).toBeResource()
    }
})

File: data/yaml_test.py
import json
import os
from pathlib import Path

import jsonschema
import yaml

CURRENT_FOLDER = Path(__file__).parent

# YAML data paths with their corresponding schema paths
resource_data_path = CURRENT_FOLDER / "resources.yml"
resource_schema_path = CURRENT_FOLDER / "resources.schema.json"

youtube_channel_data_path = CURRENT_FOLDER / "youtube_channel_ids.yml"
youtube_channel_schema_path = CURRENT_FOLDER / "youtube_channel_ids.schema.json"

resource_tags_data_path = CURRENT_FOLDER / "resource_tags.yml"
resource_tags_schema_path = CURRENT_FOLDER / "resource_tags.schema.json"


def validate_yaml(yaml_path, schema_path):
    with open(yaml_path) as f:
        data = yaml.safe_load(f)

    with open(schema_path) as f:
        schema = json.load(f)

    jsonschema.validate(data, schema)


def test_yml_against_schemas():
    validate_yaml(resource_data_path, resource_schema_path)
    validate_yaml(youtube_channel_data_path, youtube_channel_schema_path)
    validate_yaml(resource_tags_data_path, resource_tags_schema_path)


def test_resource_tags():
    """
    Reads in valid resource tags from `resource_tags_data_path` and checks that
    all tags in `resource_data_path` are valid.

    Also verifys that the tags in `resource_tags_data_path` are unique.
    """
    with open(resource_tags_data_path) as f:
        tag_data = yaml.safe_load(f)
        valid_tags = [tag["name"] for tag in tag_data]

    with open(resource_data_path) as f:
        resource_data = yaml.safe_load(f)

    # Checking for duplicate tags in tag data file
    if len(valid_tags) != len(set(valid_tags)):
        raise Exception("Duplicate tag names in resource_tags.yml")

    # Go through resource data and check tags
    errors = []
    for resource in resource_data:
        for tag in resource["tags"]:
            if tag not in valid_tags:
                errors.append(
                    f"Invalid tag {tag!r} in resource {resource['title']!r} "
                    f"({resource['url']})"
                )

    if errors:
        raise Exception("\n".join(errors))
